# Demonstrating the relationship between sample size and the accuracy of the sample mean

In [4]:
import numpy as np

In [48]:
!pip install plotly

  Obtaining dependency information for plotly from https://files.pythonhosted.org/packages/f9/14/abe5ce876ab5b66ee3c691bf537fcd43d037aea55d447aacf74630a8f31e/plotly-6.8.0-py3-none-any.whl.metadata
  Using cached plotly-6.8.0-py3-none-any.whl.metadata (9.0 kB)
  Obtaining dependency information for narwhals>=1.15.1 from https://files.pythonhosted.org/packages/48/ca/36339329c4604adbcc99c899b7eb1ce1a555c499b6a6860757dc9bfed36d/narwhals-2.22.1-py3-none-any.whl.metadata
  Using cached narwhals-2.22.1-py3-none-any.whl.metadata (15 kB)
Using cached plotly-6.8.0-py3-none-any.whl (9.9 MB)
Using cached narwhals-2.22.1-py3-none-any.whl (454 kB)


### Creates a starting population of 100,000 numbers and calculates the mean of the population.

In [111]:
np.random.seed(62)
population = np.random.normal(100, 40, 100000)
print("Population mean:", population.mean())

Population mean: 100.10416601326479


## Takes a random sample of 50,000 values from the population and checks mean.

### Mean of 50,000 (99.97) is 0.14 away from population mean (100.10)

In [112]:
sample_50000 = np.random.choice(population, 50000)
print("Sample of 50,000 mean:", sample_50000.mean())

Sample of 50,000 mean: 99.96665016314387


### Mean of 25,000 (99.78) is 0.33 away from population mean (100.10)

In [113]:
sample_25000 = np.random.choice(population, 25000)
print("Sample of 25,000 mean:",sample_25000.mean())

Sample of 25,000 mean: 99.77781162278593


### Mean of 12,500 (99.86) is 0.24 away from population mean (100.10)

In [114]:
sample_12500 = np.random.choice(population, 12500)
print("Sample of 12,500 mean:", sample_12500.mean())

Sample of 12,500 mean: 99.86096828959501


### Mean of 6,250 (100.35) is 0.24 away from population mean (100.10)

In [115]:
sample_6250 = np.random.choice(population, 6250)
print("Sample of 6,250 mean:", sample_6250.mean())

Sample of 6,250 mean: 100.34573754570684


### Mean of 3,125 (99.38) is 0.73 away from population mean (100.10)

In [116]:
sample_3125 = np.random.choice(population, 3125)
print("Sample of 3,125 mean:", sample_3125.mean())

Sample of 3,125 mean: 99.37521022108685


### Mean of 1,000 (99.21) is 0.90 away from population mean (100.10)

In [117]:
sample_1000 = np.random.choice(population, 1000)
print("Sample of 1,000 mean:", sample_1000.mean())

Sample of 1,000 mean: 99.20561878691515


### Mean of 500 (97.10) is 3.01 away from population mean (100.10)

In [118]:
sample_500 = np.random.choice(population, 500)
print("Sample of 500 mean:", sample_500.mean())

Sample of 500 mean: 97.09667763288356


## Summary -- As the sample size decreases from 50,000 down to 500, the sample mean becomes less reliable and gets further away from the original true population mean of 100.10.




# 
# Demonstrating the difference between mean, median, mode, variance, standard deviation, uniform distribution, and normal distribution.

## Imports packages, then sends a request to the OpenIntro website to download the oscars.csv file

In [49]:
import pandas as pd
import plotly.express as px
import plotly.io as pio
pio.renderers.default = "iframe"
import requests
from io import StringIO

url = "https://www.openintro.org/data/csv/oscars.csv"
headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3'}
response = requests.get(url, headers=headers)
response.raise_for_status()
df = pd.read_csv(StringIO(response.text))

## df is the dataframe with the Oscar awards data taken from openinto data sets.

#### This dataset contains variables - winner's name, age, birthplace, birth date, and the movie they won for.
#### We are going to use the "age" column to demonstrate mean, median, mode, variance, standard deviation, and compare it against simulated uniform and normal distributions

In [50]:
df

,oscar_no,oscar_yr,award,name,movie,age,birth_pl,birth_date,birth_mo,birth_d,birth_y
0,1,1929,Best actress,Janet Gaynor,7th Heaven,22,Pennsylvania,1906-10-06,10,6,1906
1,2,1930,Best actress,Mary Pickford,Coquette,37,Canada,1892-04-08,4,8,1892
2,3,1931,Best actress,Norma Shearer,The Divorcee,28,Canada,1902-08-10,8,10,1902
3,4,1932,Best actress,Marie Dressler,Min and Bill,63,Canada,1868-11-09,11,9,1868
4,5,1933,Best actress,Helen Hayes,The Sin of Madelon Claudet,32,Washington DC,1900-10-10,10,10,1900
...,...,...,...,...,...,...,...,...,...,...,...
179,87,2015,Best actor,Eddie Redmayne,The Theory of Everything,32,England,1982-01-06,1,6,1982
180,88,2016,Best actor,Leonardo Di Caprio,The Revenant,41,California,1974-11-11,11,11,1974
181,89,2017,Best actor,Casey Affleck,Manchester by the Sa,41,Massachusetts,1975-08-12,8,12,1975
182,90,2018,Best actor,Gary Oldman,Darkest Hour,59,England,1958-03-21,3,21,1958


### Mean - the mean is the average value in the data

#### Uses pandas .mean() function to calculate the mean age from the dataset.

In [73]:
mean_age = df["age"].mean()
print("The mean age of an Oscar Acting Winner is", round(mean_age, 2))

The mean age of an Oscar Acting Winner is 40.05


### Mean - the median is the middle value for a data set in order

#### Uses pandas .median() function to calculate the median age from the dataset.

In [76]:
median_age = df["age"].median()
print("The median age of an Oscar Acting Winner is", median_age)

The median age of an Oscar Acting Winner is 38.0


### Mode - the mode is value in the data set that appears the most frequently

#### Uses pandas .mode() function to get the most frequent age values from the dataset

In [63]:
mode_age = df["age"].mode()[0]
mode_age

41

### Variance - the variance is the measure of the spread of a data set relative to its mean

#### Uses pandas .var() function to get the variance of the "age" column

In [54]:
variance = df["age"].var()
variance

123.7079769541459

### Standard Deviation - the standard deviation is the average distance between data values and the mean

#### Uses pandas .std() function to calculate the standard deviation of age from the dataset.

In [55]:
standard_dev = df["age"].std()
standard_dev

11.122408774817886

### Uniform Distribution - distribution where the data is spread roughly symmetric and constant

#### Creates a histogram with a random distribution with 1000 random values to simulate a uniform distribution.

In [120]:
uniform_distribution = np.random.uniform(low=df["age"].min(), high=df["age"].max(), size=1000)

fig = px.histogram(uniform_distribution, nbins=20, title="Uniform Distribution")
fig.show()

### Normal Distribution - distribution where the data is symmetric and centered around the mean.

#### Creates a histogram with a random distribution with 1000 random values to simulate a normal distribution.

In [119]:
normal_distribution = np.random.normal(loc=mean_age, scale=standard_dev, size=1000)

fig = px.histogram(normal_distribution, nbins=20, title="Normal Distribution")
fig.show()